contributors/design-proposals/storage/pv-monitoring-proposal.md

```diff
@@ -0,0 +1,128 @@
+
+## User Experience
+### Use Cases
+
+* Users create many local PVs manually and want to monitor their health condition;
+* Users want to know if the volumes their apps are referencing are healthy or not;
+* If node breaks down, we need to reschedule pods referencing that node's local PVs (the application has data backup and can restore it or can tolerate data loss);
+
+## Implementation
+
+As mentioned above, we can split this into two parts and put them in the external repo at first.
+
+### Monitoring controller: 
+
+Like PV controller, monitoring controller should check PVs’ health condition periodically and taint them if PVs are unhealthy.
+Health checking implementation should be per plugin. Each volume plugin needs to have its own methods to check its volumes.
+
+#### For in-tree volume plugins(except local storage):
+
+We can add a new volume plugin interface: StatusCheckingVolumePlugin.
+```
+  type StatusCheckingVolumePlugin interface {
+       VolumePlugin
+
+      CheckStatus(spec *Spec) (string, error)
+  }
+```
+And each volume plugin will implement it. The entire monitoring controller workflow is:
+
+* Fill PV cache with initial data from etcd
+* Resync and check volumes status periodically
+* Taint PV if the volume status is abnormal
+* Remove the taints if the volume becomes normal again
+
+#### For out-tree volume plugins(except local storage):
+
+We can create a new controller called MonitorController like ProvisionController, 
+which is responsible for creating informers, watching Node and PV events and calling each plugin’s monitor functions. 
+And each volume plugin will create its own monitor to check its volumes’ status.
+
+#### For local storage:
+
+The local storage PV monitor consists of two parts
+
+* create a daemonset on every node, which is responsible for monitoring local PVs in that specific node, no matter the PVs are created manually or by provisioner;
+* create a monitor controller, which is responsible for watching PVs and Nodes events. PVs may be updated if they are unhealthy and we also need to react to node failure event.
+
+At the first phase, we can support local storage monitoring first.
+
+Take local storage as an example, detailed monitoring implementation may be like this:
+
+```
+ func (mon *minitor) CheckStatus(spec *v1.PersistentVolumeSpec) (string, error) {
+        // check if PV is local storage
+        if pv.Spec.Local == nil { 
+                glog.Infof("PV: %s is not local storage", pv.Name)
+                return ...
+        }
+        
+        // check if PV belongs to this node
+        if checkNodeAffinity(...) {
+                glog.Infof("PV: %s does not belong to this node", pv.Name)
+                return ...
+        }
+
+        // check volume health condition depending on device type
+        ...
+
+}
+```
+If monitor finds that one PV is unhealthy, it will mark the PV by adding annotations including timestamp. 
```

Comment:
Can you define the annotation key/value?

Reply:
yeah, sure, defined in demo PR: kubernetes-incubator/external-storage#528
will update this proposal

Reply:
Can you define the annotation format and syntax in this design doc, and also put some of the examples here too?

Reply:
done